h2/stream.py

```diff
@@ -279,19 +315,40 @@ def send_reset(self, previous_state):
         it's the first time we've done this if the stream is currently in a
         state other than CLOSED.
         """
-        events = []
+        assert previous_state != StreamState.CLOSED
 
-        if previous_state != StreamState.CLOSED:
-            event = StreamReset()
-            event.stream_id = self.stream_id
-            event.error_code = ErrorCodes.STREAM_CLOSED
-            event.remote_reset = False
-            events.append(event)
+        self.stream_closed_by = StreamClosedBy.SEND_RST_STREAM
 
         error = StreamClosedError(self.stream_id)
-        error._events = events
+
+        event = StreamReset()
+        event.stream_id = self.stream_id
+        event.error_code = ErrorCodes.STREAM_CLOSED
+        event.remote_reset = False
+        error._events = [event]
         raise error
 
+    def recv_on_closed_stream(self, previous_state):
+        """
+        Called when an unexpected frame is received on an already-closed
+        stream.
+
+        If the stream is closed by RST_STREAM, it will raise StreamClosedError
+        to reset stream. If the stream is closed by END_STREAM, it will raise
+        ProtocolError and close the connection.
+        """
+        assert previous_state == StreamState.CLOSED
+        assert self.stream_closed_by is not None
+
+        if self.stream_closed_by in (StreamClosedBy.RECV_RST_STREAM,
+                                     StreamClosedBy.SEND_RST_STREAM):
+            raise StreamClosedError(self.stream_id)
+        else:
+            raise ProtocolError(
```

Comment:
I'm not sure that it's worth or not to add assert here.
I'm fine for with assert here or not!